python/cuml/cluster/agglomerative.pyx

```diff
@@ -0,0 +1,249 @@
+        which distance to use between sets of observations. The algorithm
+        will merge the pairs of clusters that minimize this criterion.
+        - 'single' uses the minimum of the distances between all
+          observations of the two sets.
+    n_neighbors : int (default = 15)
+        The number of neighbors to compute when connectivity = "knn"
+    connectivity : {"pairwise", "knn"}, (default = "knn")
+        The type of connectivity matrix to compute.
+        - 'pairwise' will compute the entire fully-connected graph of
+          pairwise distances between each set of points. This is the
+          fastest to compute and can be very fast for smaller datasets
+          but requires O(n^2) space.
+        - 'knn' will sparsify the fully-connected connectivity matrix to
+          save memory and enable much larger inputs. "n_neighbors" will
+          control the amount of memory used and the graph will be connected
+          automatically in the event "n_neighbors" was not large enough
+          to connect it.
+    output_type : {'input', 'cudf', 'cupy', 'numpy', 'numba'}, default=None
+        Variable to control output type of the results and attributes of
+        the estimator. If None, it'll inherit the output type set at the
+        module level, `cuml.global_settings.output_type`.
+        See :ref:`output-data-type-configuration` for more info.
+    """
+
+    labels_ = CumlArrayDescriptor()
+    children_ = CumlArrayDescriptor()
+
+    def __init__(self, n_clusters=2, affinity="euclidean", linkage="single",
+                 handle=None, verbose=False, connectivity='knn',
+                 n_neighbors=10, output_type=None):
+
+        super(AgglomerativeClustering, self).__init__(handle,
+                                                      verbose,
+                                                      output_type)
+
+        if linkage is not "single":
+            raise ValueError("Only single linkage clustering is "
+                             "supported currently")
+
+        if connectivity not in ["knn", "pairwise"]:
+            raise ValueError("'connectivity' can only be one of "
+                             "{'knn', 'pairwise'}")
+
+        if n_clusters <= 0:
+            raise ValueError("'n_clusters' must be >= 1")
+
+        if n_neighbors > 1023 or n_neighbors < 2:
+            raise ValueError("'n_neighbors' must be a positive number "
+                             "between 2 and 1023")
+
+        if affinity not in _metrics_mapping:
+            raise ValueError("'affinity' %s is not supported." % affinity)
+
+        self.n_clusters = n_clusters
+        self.affinity = affinity
+        self.linkage = linkage
+        self.n_neighbors = n_neighbors
+        self.connectivity = connectivity
+
+        self.labels_ = None
+        self.n_clusters_ = None
+        self.n_leaves_ = None
+        self.n_connected_components_ = None
+        self.children_ = None
+        self.distances_ = None
```

Comment:
```suggestion
        self.n_clusters_ = None
        self.n_leaves_ = None
        self.n_connected_components_ = None
        self.distances_ = None
```

They are not needed so I would recommend removing the lines, but its just a minor nitpick